docs: add reusable parts (#1928)
asyncapi/website: pages/docs/concepts/asyncapi-document/resuable-parts.md

---
title: Reusable parts
weight: 280
---

> Reusable parts in AsyncAPI provide flexibility, modularity, and code reusability. Using reusable parts makes it easier to generate code and validate the specifications. You can reuse specific document sections such as Messages or schema definitions.

Reviewer: > Using reusable parts makes it easier to generate code and validate the specifications.

What do you mean here?

Reusable parts allow you to split up the AsyncAPI document into many files and reference them using the Reference Object. You can use the $ref keyword to reference the same or another local file or external URL.

## Same file

You can use the $ref keyword to reference a component within the same document. The diagram below defines referencing a component within the same document.

```mermaid
graph LR
style A fill:#47BCEE, stroke:#333, stroke-width:2px;
style C fill:#47BCEE, stroke:#333, stroke-width:2px;
style D fill:#47BCEE, stroke:#333, stroke-width:2px;
style B fill:#47BCEE, stroke:#333, stroke-width:2px;

A[MyMessageSchema]
B[components]
C[myChannel]
D[$ref keyword]

C -->|publishes| D
D -->|references| B

> B -->|references| A

Reviewer: I don't fully understand the meaning of this diagram. What are you trying to explain here?

```

In the example below, you define a component called MyMessageSchema under the schemas section to describe the structure of a message. Under the publish operation of myChannel, you reference the MyMessageSchema component using the $ref keyword.

```yaml
channels:
  myChannel:
    publish:
      message:

>         $ref: '#/components/schemas/MyMessageSchema'

Reviewer: ```suggestion
        payload:
          $ref: '#/components/schemas/MyMessageSchema'
```

components:
  schemas:
    MyMessageSchema:
      type: object
      properties:
        message:
          type: string
```

## Another local document

You can use the $ref keyword to reference another local document. Ensure the path to the local file is correct and accessible from your main AsyncAPI document.

The diagram below defines the process of referencing another local document.

```mermaid
graph LR
style A fill:#47BCEE, stroke:#333, stroke-width:2px;
style C fill:#47BCEE, stroke:#333, stroke-width:2px;
style D fill:#47BCEE, stroke:#333, stroke-width:2px;
style B fill:#47BCEE, stroke:#333, stroke-width:2px;

A[Main AsyncAPI Document]
B[message-schema.yaml]
C[components/schemas]
D[MyMessageSchema]

A -->|references| B
B -->|contains| C

> C -->|defines| D

Reviewer: Same as with the other diagram. It's not clear to me what's the intent or meaning.

```

In the example below, you reference the component from the message-schema.yaml.

```yaml
components:
  messages:
    UserSignup:
      name: UserSignup
      title: User signup
      summary: Action to sign a user up.
      description: A longer description
      contentType: application/json
      payload: null
```

To use message-schema.yaml in another local document such as consume-schema.yaml

```yaml
channels:
  user/signedup:
    publish:
      message:
        $ref: ./message-schema.yaml#/components/messages/UserSignup

> ```

Reviewer: Which example belongs to which file? It's not clear to me here 🤔

## External URL

You can use the $ref keyword to reference an external URL. Ensure the external URL should provide the referenced component in a compatible format, such as YAML or JSON.

The diagram below defines the process of referencing an external URL.

```mermaid
graph LR
style A fill:#47BCEE, stroke:#333, stroke-width:2px;
style C fill:#47BCEE, stroke:#333, stroke-width:2px;
style D fill:#47BCEE, stroke:#333, stroke-width:2px;
style B fill:#47BCEE, stroke:#333, stroke-width:2px;

A[AsyncAPI Document]
B[https://example.com/my-component.yaml]
C[components/schemas]
D[MyComponent]

A -->|references| B
B -->|contains| C

> C -->|defines| D

Reviewer: Same as with the other diagrams.

---
title: Reusable parts

> weight: 249

Reviewer 2: Either this page weight number is not sufficiently high, or the weight in the `_section.md` needs to be fixed... do you notice in your preview link that the `Overview` page shows incorrectly AFTER this new document?

BTW, moving forward, let's remember to always review your work via the generated Netlify preview link generated in each PR. Ideally, you should have noticed this before I did 😄

<img width="400" alt="Screen Shot 2023-08-31 at 4 56 35 PM" src="https://github.com/asyncapi/website/assets/19964402/4ef6631c-6e1c-48d3-bbba-4ff1aec0f3c8">

---

Reusable parts in AsyncAPI provide flexibility, modularity, and code reusability. Using reusable parts makes it easier to generate code and validate the specifications. You can reuse specific document sections such as Messages or schema definitions.

> Reusable parts allow you to split up the AsyncAPI document into many files and reference them using the Reference Object ($ref). You can use the $ref keyword to reference the same or another local file or external URL.

Reviewer 2: Sometimes you write `$ref` and sometimes only `ref`. For consistency, you must stick to one. Please update all these mentions consistently. ✌🏽

> Reusable parts allow you to split up the AsyncAPI document into many files and reference them using the Reference Object ($ref). You can use the $ref keyword to reference the same or another local file or external URL.

Reviewer 2: In line 8, you wrote `($ref)`, but this seems to be the only place where you added parenthesis. If you review your work on the Netlify preview link, you'll notice that adding parenthesis created a different markup style where it applies a grey highlight over the text.

To me, it seems you need to **remove the parenthesis** in that line in order to be consistent in all mentions.

<img width="400" alt="Screen Shot 2023-08-31 at 5 02 54 PM" src="https://github.com/asyncapi/website/assets/19964402/cc7b43af-d122-48ee-a6ae-45376ea3a000">

Author: I added the parenthesis in the first paragraph for our reader to know, Reference Object means $ref

## Same file

> You can use the ref keyword to reference a component within the same document. The image below defines referencing a component within the same document.

Reviewer 2: I notice you're using the word `image` to introduce the diagrams in this PR, but we should be consistent across the documentation. Currently, we are using the word `diagram`. 

Please update all your mentions accordingly. 😄

```mermaid
graph LR
style A fill:#47BCEE, stroke:#333, stroke-width:2px;
style C fill:#47BCEE, stroke:#333, stroke-width:2px;
style D fill:#47BCEE, stroke:#333, stroke-width:2px;
style B fill:#47BCEE, stroke:#333, stroke-width:2px;

A[MyMessageSchema]
B[components]
C[myChannel]
D[$ref keyword]

C -->|publishes| D
D -->|references| B
B -->|references| A
```

> In the example below, you define a component called MyMessageSchema under the schemas section to describe the structure of a message. Under the publish operation of myChannel, you reference the MyMessageSchema component using the $ref keyword.

Reviewer 2: Let's apply styling consistently to code objects/items here, such as `MyMessageSchema`, `myChannel`, etc. 

I notice you did so on other PRs. Can you please update this PR accordingly? 😄

```yaml
channels:
  myChannel:
    publish:
      message:
        $ref: '#/components/schemas/MyMessageSchema'
components:
  schemas:
    MyMessageSchema:
      type: object
      properties:
        message:
          type: string
```

## Another local document

You can use the ref keyword to reference another local document. Ensure the path to the local file is correct and accessible from your main AsyncAPI document.

The image below defines the process of referencing another local document.

```mermaid
graph LR
style A fill:#47BCEE, stroke:#333, stroke-width:2px;
style C fill:#47BCEE, stroke:#333, stroke-width:2px;
style D fill:#47BCEE, stroke:#333, stroke-width:2px;
style B fill:#47BCEE, stroke:#333, stroke-width:2px;

A[Main AsyncAPI Document]
B[message-schema.yaml]
C[components/schemas]
D[MyMessageSchema]

A -->|references| B
B -->|contains| C
C -->|defines| D
```

> In the example below, you reference the component from the message-schema.yaml file.

Reviewer 2: ```suggestion
In the example below, you reference the component from the `message-schema.yaml` file.
```

```yaml

> ##### ./message-schema.yaml

Reviewer 2: Why the need for `#####` included in this line? Seems like a mistake.

Author: We need to indicate file names, since we are using a couple of files.

components:
  messages:
    UserSignup:
      name: UserSignup
      title: User signup
      summary: Action to sign a user up.
      description: A longer description
      contentType: application/json
      payload: null
```

```yaml

> ##### ./asyncapi.B.yaml

Reviewer 2: Why the need for `#####` included in this line? Seems like a mistake.

Author: We need to indicate file names, since we are using a couple of files.

channels:
  user/signedup:
    publish:
      message:

>         $ref: ./asyncapi.A.yaml#/components/messages/userSignedUp

Reviewer 3: Should be `UserSignup`

```

## External URL

> You can use the ref keyword to reference an external URL. Ensure the external URL should provide the referenced component in a compatible format, such as YAML or JSON.

Reviewer 2: ```suggestion
You can use the ref keyword to reference an external URL. Ensure the external URL provides the referenced component in a compatible format, such as YAML or JSON.
```

The image below defines the process of referencing an external URL.

```mermaid
graph LR
style A fill:#47BCEE, stroke:#333, stroke-width:2px;
style C fill:#47BCEE, stroke:#333, stroke-width:2px;
style D fill:#47BCEE, stroke:#333, stroke-width:2px;
style B fill:#47BCEE, stroke:#333, stroke-width:2px;

A[AsyncAPI Document]
B[https://example.com/my-component.yaml]
C[components/schemas]
D[MyComponent]

A -->|references| B
B -->|contains| C
C -->|defines| D
```

In the example below, you reference the component from an external URL. The $ref value specifies the full URL to the external resource and the component's location.

```yaml

> ##### ./asyncapi.A.yaml

Reviewer 2: Same question as line 88... Why the need for `#####` included in this line? Seems like a mistake. 😄

Author: We need to indicate file names, since we are using a couple of files.
